src/io/frame.rs

```diff
@@ -357,27 +357,49 @@ impl<T: Io, C: Codec> Sink for Framed<T, C> {
     type SinkError = io::Error;
 
     fn start_send(&mut self, item: C::Out) -> StartSend<C::Out, io::Error> {
+        // If the buffer is already over 8KiB, then attempt to flush it. If after flushing it's
+        // *still* over 8KiB, then apply backpressure (reject the send).
+        if self.wr.len() > 8 * 1024 {
+            try!(self.poll_complete());
+            if self.wr.len() > 8 * 1024 {
+                return Ok(AsyncSink::NotReady(item));
+            }
+        }
+
         self.codec.encode(item, &mut self.wr);
         Ok(AsyncSink::Ready)
     }
 
     fn poll_complete(&mut self) -> Poll<(), io::Error> {
-        // Try flushing the underlying IO
-        try_nb!(self.upstream.flush());
-
         trace!("flushing framed transport");
 
-        loop {
-            if self.wr.len() == 0 {
-                trace!("framed transport flushed");
-                return Ok(Async::Ready(()));
+        let mut offset = 0;
+        while self.wr.len() - offset > 0 {
+            trace!("writing; remaining={}", self.wr.len() - offset);
+            match self.upstream.write(&self.wr[offset..]) {
+                Ok(0) => {
+                    self.wr.drain(..offset);
+                    return Err(io::Error::new(io::ErrorKind::WriteZero,
+                                              "failed to write frame to transport"));
+                },
+                Ok(n) => offset += n,
+                Err(error) => {
+                    self.wr.drain(..offset);
+                    if error.kind() == io::ErrorKind::WouldBlock {
+                        return Ok(Async::NotReady);
+                    } else {
+                        return Err(error);
+                    }
+                }
             }
+        }
+        self.wr.drain(..offset);
```

Comment:
The major difference between this loop and the previous loop seems to be deferring the call to `drain` until the very end, which seems to be an optimization? Initial testing I've done ended up never showing that improvement in profiles, so I'm curious if you were able to get speedups with a change like this?